src/runtime/cache.cpp

```diff
@@ -562,6 +571,41 @@ WEAK void halide_memoization_cache_cleanup() {
     least_recently_used = nullptr;
 }
 
+WEAK void halide_memoization_cache_evict(void *user_context, uint64_t eviction_key) {
+    ScopedMutexLock lock(&memoization_lock);
+
+    for (size_t i = 0; i < kHashTableSize; i++) {
+        CacheEntry *entry = cache_entries[i];
+        if (entry != nullptr) {
+            CacheEntry **prev = &cache_entries[i];
+            while (entry != nullptr) {
+                CacheEntry *next = entry->next;
+                if (entry->has_eviction_key && entry->eviction_key == eviction_key) {
```

Comment:
Can this list removal be in a helper function? It looks like all the other potential uses are removing from the end and so are a bit simpler, but I think maybe it would be better if they all used a more general list removal helper anyways?